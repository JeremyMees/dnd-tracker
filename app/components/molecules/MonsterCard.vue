<script setup lang="ts">
import { abilities } from '~~/constants/dnd-rules'

defineEmits<{ add: [monster: Open5eItem] }>()

withDefaults(defineProps<{
  monster: Open5eItem
  addable?: boolean
}>(), {
  addable: false,
})

const abilitiesNames = ref<(keyof Open5eItem)[]>(['strength', 'dexterity', 'constitution', 'intelligence', 'wisdom', 'charisma'])
const stats = ref<(keyof Open5eItem)[]>(['type', 'subtype', 'size', 'alignment', 'xp'])
const isOpen = ref<boolean>(false)
</script>

<template>
  <UiCard
    class="relative bg-background/50 border-4 border-background"
    @dblclick="isOpen = !isOpen"
  >
    <UiCardHeader>
      <button
        v-if="addable"
        v-tippy="{
          content: $t('actions.add'),
          placement: 'left',
        }"
        :aria-label="$t('actions.add')"
        class="absolute right-4 top-4 icon-btn-success"
        @click="$emit('add', monster)"
      >
        <Icon
          name="tabler:plus"
          class="size-4 min-w-4"
          aria-hidden="true"
        />
      </button>
      <UiCardTitle>{{ monster.name }}</UiCardTitle>
      <div class="flex gap-4">
        <div
          v-tippy="{ content: 'CR' }"
          class="flex gap-1"
        >
          <Icon
            name="tabler:skull"
            class="w-6 h-6 text-warning"
            aria-hidden="true"
          />
          <p class="font-bold">
            {{ monster.challenge_rating || '_' }}
          </p>
        </div>
        <div
          v-tippy="{ content: 'AC' }"
          class="flex gap-1"
        >
          <Icon
            name="tabler:shield"
            class="w-6 h-6 text-help"
            aria-hidden="true"
          />
          <p class="font-bold">
            {{ monster.armor_class || '_' }}
          </p>
        </div>
        <div
          v-tippy="{ content: 'HP' }"
          class="flex gap-1"
        >
          <Icon
            name="tabler:heart"
            class="w-6 h-6 text-destructive"
            aria-hidden="true"
          />
          <p class="font-bold">
            {{ monster.hit_points || '_' }}
          </p>
        </div>
      </div>
    </UiCardHeader>
    <UiCardContent>
      <div class="flex gap-x-4 gap-y-1 flex-wrap">
        <div
          v-for="(ability, index) in abilities"
          :key="ability"
          class="flex gap-1"
        >
          <span class="text-muted-foreground">
            {{ ability }}:
          </span>
          <span>
            {{ abilitiesNames[index] ? monster[abilitiesNames[index]] : '_' }}
          </span>
        </div>
      </div>
      <div class="flex gap-x-4 gap-y-1 flex-wrap">
        <template
          v-for="stat in stats"
          :key="stat"
        >
          <div
            v-if="monster[stat]"
            class="flex gap-1"
          >
            <span class="text-muted-foreground">
              {{ stat }}:
            </span>
            <span class="lowercase">
              {{ monster[stat] || '_' }}
            </span>
          </div>
        </template>
      </div>
      <ActionsTable
        v-if="monster.actions && isOpen"
        :actions="monster.actions"
        :legendary-actions="monster.legendary_actions"
        :special-abilities="monster.special_abilities"
        :reactions="monster.reactions"
        :class="{ 'py-5': isOpen }"
      />
    </UiCardContent>
    <UiCardFooter>
      <div class="flex justify-end w-full">
        <button
          class="flex gap-2 text-foreground"
          :aria-label="$t(`actions.read${isOpen ? 'Less' : 'More'}`)"
          @click="isOpen = !isOpen"
        >
          <p>
            {{ $t(`actions.read${isOpen ? 'Less' : 'More'}`) }}
          </p>
          <Icon
            name="tabler:chevron-down"
            class="duration-200 size-6 stroke-2"
            :class="{ 'rotate-180': isOpen }"
            aria-hidden="true"
          />
        </button>
      </div>
    </UiCardFooter>
  </UiCard>
</template>
