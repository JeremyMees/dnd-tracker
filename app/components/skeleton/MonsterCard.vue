<template>
  <UiCardBase class="bg-background/50 border-4 border-background">
    <UiCardHeader>
      <UiCardTitle>
        <UiSkeletonBase class="w-[120px] h-7" />
      </UiCardTitle>
      <div class="flex gap-4 mt-1">
        <UiSkeletonBase class="w-10 h-6" />
        <UiSkeletonBase class="w-10 h-6" />
        <UiSkeletonBase class="w-10 h-6" />
      </div>
    </UiCardHeader>
    <UiCardContent>
      <UiSkeletonBase class="w-full h-16" />
    </UiCardContent>
    <UiCardFooter>
      <div class="flex justify-end w-full">
        <UiSkeletonBase class="w-[120px] h-7" />
      </div>
    </UiCardFooter>
  </UiCardBase>
</template>
