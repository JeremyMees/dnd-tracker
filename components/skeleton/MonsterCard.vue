<template>
  <UiCard class="bg-background/50 border-4 border-background">
    <UiCardHeader>
      <UiCardTitle>
        <UiSkeleton class="w-[120px] h-7" />
      </UiCardTitle>
      <div class="flex gap-4 mt-1">
        <UiSkeleton class="w-10 h-6" />
        <UiSkeleton class="w-10 h-6" />
        <UiSkeleton class="w-10 h-6" />
      </div>
    </UiCardHeader>
    <UiCardContent>
      <UiSkeleton class="w-full h-16" />
    </UiCardContent>
    <UiCardFooter>
      <div class="flex justify-end w-full">
        <UiSkeleton class="w-[120px] h-7" />
      </div>
    </UiCardFooter>
  </UiCard>
</template>
