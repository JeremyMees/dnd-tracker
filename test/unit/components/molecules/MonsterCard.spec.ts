import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it } from 'vitest'
import MonsterCard from '~/components/molecules/MonsterCard.vue'
import open5eItem from '~~/test/unit/fixtures/open5e-item.json'

interface Props {
  monster: Open5eItem
  addable?: boolean
}

const props: Props = {
  monster: open5eItem as unknown as Open5eItem,
  addable: false,
}

describe('MonsterCard', async () => {
  it('Should match snapshot', async () => {
    const component = await mountSuspended(MonsterCard, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should render correct with default props', async () => {
    const component = await mountSuspended(MonsterCard, { props })

    expect(component.find('[data-test-title]').text()).toBe(props.monster.name)
    expect(component.find('[data-test-add-button]').exists()).toBe(props.addable)
    expect(component.find('[data-test-actions-table]').exists()).toBeFalsy()
    expect(component.find('[data-test-expand-button]').exists()).toBeTruthy()
  })

  it('Should be able to expand the card', async () => {
    const component = await mountSuspended(MonsterCard, { props })
    const button = component.find('[data-test-expand-button]')

    expect(component.find('[data-test-actions-table]').exists()).toBeFalsy()

    await button.trigger('click')
    await nextTick()

    expect(component.find('[data-test-actions-table]').exists()).toBeTruthy()
  })

  it('Should be able to add the monster', async () => {
    const component = await mountSuspended(MonsterCard, { props: { ...props, addable: true } })
    const button = component.find('[data-test-add-button]')

    expect(button.exists()).toBeTruthy()

    await button.trigger('click')
    await nextTick()

    expect(component.emitted('add')).toBeTruthy()
    expect(component.emitted('add')![0]).toEqual([props.monster])
  })
})
