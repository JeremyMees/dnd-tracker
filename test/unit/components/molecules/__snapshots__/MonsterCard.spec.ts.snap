// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MonsterCard > Should match snapshot 1`] = `
"<div class="rounded-lg text-card-foreground shadow-sm relative bg-background/50 border-4 border-background">
  <div class="flex flex-col gap-y-1.5 p-6">
    <!--v-if-->
    <h3 class="text-2xl font-semibold leading-none tracking-tight" data-test-title="">Aboleth</h3>
    <div class="flex gap-4">
      <div class="flex gap-1"><span class="iconify i-tabler:skull w-6 h-6 text-warning" aria-hidden="true"></span>
        <p class="font-bold">10</p>
      </div>
      <div class="flex gap-1"><span class="iconify i-tabler:shield w-6 h-6 text-help" aria-hidden="true"></span>
        <p class="font-bold">17</p>
      </div>
      <div class="flex gap-1"><span class="iconify i-tabler:heart w-6 h-6 text-destructive" aria-hidden="true"></span>
        <p class="font-bold">135</p>
      </div>
    </div>
  </div>
  <div class="p-6 pt-0">
    <div class="flex gap-x-4 gap-y-1 flex-wrap">
      <div class="flex gap-1"><span class="text-muted-foreground">STR: </span><span>21</span></div>
      <div class="flex gap-1"><span class="text-muted-foreground">DEX: </span><span>9</span></div>
      <div class="flex gap-1"><span class="text-muted-foreground">CON: </span><span>15</span></div>
      <div class="flex gap-1"><span class="text-muted-foreground">INT: </span><span>18</span></div>
      <div class="flex gap-1"><span class="text-muted-foreground">WIS: </span><span>15</span></div>
      <div class="flex gap-1"><span class="text-muted-foreground">CHA: </span><span>18</span></div>
    </div>
    <div class="flex gap-x-4 gap-y-1 flex-wrap">
      <div class="flex gap-1"><span class="text-muted-foreground">type: </span><span class="lowercase">Aberration</span></div>
      <!--v-if-->
      <div class="flex gap-1"><span class="text-muted-foreground">size: </span><span class="lowercase">Large</span></div>
      <div class="flex gap-1"><span class="text-muted-foreground">alignment: </span><span class="lowercase">lawful evil</span></div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class="flex items-center p-6 pt-0">
    <div class="flex justify-end w-full"><button data-test-expand-button="" class="flex gap-2 text-foreground" aria-label="actions.readMore">
        <p>actions.readMore</p><span class="iconify i-tabler:chevron-down duration-200 size-6 stroke-2" aria-hidden="true"></span>
      </button></div>
  </div>
</div>"
`;
